### ISSUE TITLE
I want to have a method and property with the same name

### ISSUE THREAD

**Issue by aymericbouzy** [Automattic/mongoose#5657]
**Do you want to request a *feature* or report a *bug*?**
feature
**What is the current behavior?**
error:  You have a method and a property in your schema both named "worker".
**Steps to reproduce.**
```js
const schema = new Schema({
  worker: {
    workerId: String,
    // some more fields
  },
  // some more fields
})

schema.methods = {
  worker() {
    return this.model("Worker").findOne({ _id: this.worker.workerId })
  },
}

const Contract = mongoose.model("Contract", schema)
```
**What is the expected behavior?**

The schema of the database was defined before I arrived. The project had not tests. Changing the schema could break everything.

I want to be able to get the worker associated to the contract by running `contract.worker()`. 

My suggestion : 
```js
const WorkerData = new Schema({
  workerId: String, 
}, {key: "worker"})

const schema = new Schema({
  workerData: WorkerData,
  // some more fields
})
```

That way, I have access to both `workerData` and `worker` names. I can define my schema the way I want it to look, and let Mongoose do the massaging to fit the existing database.

**Please mention your node.js, mongoose and MongoDB version.**
node: 8.4.0
mongoose: 4.11.11
mongoDB: 2.2.31

**Comment by aymericbouzy** [comment]
Maybe I can try using Mongoose Middleware ? Something of the kind : 
```js
schema.renameKeys = {
  workerData: "worker",
}

schema.pre("save", function(next) {
  this._savedDocument = {}
  Object.keys(this).forEach(key => {
    const renamedKey = schema.renameKeys[key]
    if (renamedKey) {
      this._savedDocument[renamedKey] = this[renamedKey]
      this[renamedKey] = this[key]
    }
  })
  next()
})

schema.post("save", function() {
  Object.keys(this._savedDocument).forEach(key => {
    this[key] = this._savedDocument[key]
  })
})
```

**Comment by wertlex** [comment]
Seems like this is not possible in JS. Consider following pure-js example:

```
const WorkerData = {
  workerId: "worker1"    // this is your property from schema
};

WorkerData.workerId = function() { ... }  // this will override original WorkrerData.workerId value
```

Same with mongoose. You can define both property and method of same name, but how to resolve this properly: `WorkerData.worker` ?

There is stackoverflow about this: https://stackoverflow.com/questions/29435891/differentiate-between-property-and-method-with-same-name

**Comment by vkarpov15** [comment]
So if you're getting this error you either added a new property named `worker` or added a new method named `worker`, which means no existing code depends on it, so the answer is just rename the newly added property/method. Not really possible to have an object property that is also a function.

**Comment by aymericbouzy** [comment]
Indeed I just added it, but some old meteor app also has access to the same database. I really cannot change the underlying database. I wanted to see if it was possible to write all my code as if the property had been named differently in the first place. So that when I have completed the migration out of the meteor app in a few months, I can finally rename the property without changing all my newly written code.
